src/machi_merkle_tree_mgr.erl

```diff
@@ -0,0 +1,335 @@
+                [F, Reason]),
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% private
+
+make_merkle_tree_mgr_name(FluName) ->
+    list_to_atom(atom_to_list(FluName) ++ "_merkle_tree_mgr").
+
+handle_load(Tid, DataDir) ->
+    Files = get_files(DataDir),
+    lists:foreach(fun(F) -> load_filename(Tid, DataDir, F) end, Files).
+
+get_files(_DataDir) ->
+    [].
+
+load_filename(Tid, DataDir, Filename) ->
+    CsumFile = machi_util:make_checksum_filename(DataDir, Filename),
+    case file:read_file(CsumFile) of
+        {error, enoent} -> 
+            insert(Tid, Filename, {0, ?MINIMUM_OFFSET, unwritten});
+        {ok, Bin} ->
+            load_bin(Tid, Filename, Bin);
+        Error ->
+            throw(Error)
+    end.
+
+load_bin(Tid, Filename, Bin) ->
+    CsumL = machi_csum_table:split_checksum_list_blob_decode(Bin),
+    iter_csum_list(Tid, Filename, CsumL).
+
+iter_csum_list(Tid, Filename, []) ->
+    insert(Tid, Filename, {0, ?MINIMUM_OFFSET, unwritten});
+iter_csum_list(Tid, Filename, L = [ {Last, _, _} | _ ]) ->
+    make_insert(Tid, Filename, Last, L).
+
+make_insert(_Tid, _Filename, _Last, []) ->
+    ok;
+%% case where Last offset matches Current, just insert current
+make_insert(Tid, Filename, Last, [H={Last, Len, _Csum}|T]) ->
+    insert(Tid, Filename, H),
+    make_insert(Tid, Filename, Last+Len, T);
+%% case where we have a hole
+make_insert(Tid, Filename, Last, [H={Off, Len, _Csum}|T]) ->
+    Hole = Off - Last,
+    insert(Tid, Filename, {Last, Hole, unwritten}),
+    insert(Tid, Filename, H),
+    make_insert(Tid, Filename, Off+Len, T).
+
+insert(Tid, Filename, {Offset, Length, trimmed}) ->
+    do_insert(Tid, Filename, Offset, Length, ?TRIMMED(Offset, Length));
+
+insert(Tid, Filename, {Offset, Length, unwritten}) ->
+    do_insert(Tid, Filename, Offset, Length, ?UNWRITTEN(Offset, Length));
+
+insert(Tid, Filename, {Offset, Length, <<_Tag:8, Csum/binary>>}) ->
+    do_insert(Tid, Filename, Offset, Length, ?WRITTEN(Offset, Length, Csum)).
+
+do_insert(Tid, Filename, Offset, Length, Csum) ->
+    MT = case find(Tid, Filename) of
+        [] ->
+            #mt{ filename = Filename };
+        V -> 
+            V
+    end,
+    ok = maybe_update(Tid, Offset, Length, Csum, MT),
+    ok.
+    
+maybe_update(Tid, Offset, Length, Csum, MT) ->
+    case orddict:find({Offset, Length}, MT#mt.leaves) of
+        error ->
+            %% range not found in our orddict, so fill it in
+            do_update(Tid, Offset, Length, Csum, MT);
+        {ok, Csum} ->
+            %% trying to insert a value we already have that
+            %% matches
+            ok;
+        {ok, ?UNWRITTEN(Offset, Length)} ->
+            %% old value was unwritten, now we are filling it in
+            %% so that's legit
+            do_update(Tid, Offset, Length, Csum, MT);
+        {ok, ?TRIMMED(Offset, Length)} ->
+            %% XXX FIXME 
+            %% Scott - range was trimmed - do we fill it in with new stuff?
```

Comment:
As I'm in the middle of trim work, for you this can leave as it is ... any byte is supposed to have three states, it's state transition is monotonic, like unwritten->written->trimmed or unwritten -> trimmed. Trimmed is a final state that goes nowhere next. If the data in flus are different, they should be repaired in that way. e.g. a moment where Flu A, B, C in a chain, Flu A has trimmed, Flu B is written, Flu C is unwritten: all Flus should have that byte trimmed eventually (or ASAP in read time).